### PR TITLE
Use ActiveSupport Hash#assert_valid_keys instead of refinement

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Refinements are problematic in a few ways:
- lack of Sorbet [support](https://github.com/sorbet/sorbet/issues/2690#issuecomment-591153351)
- uncertainty around what methods are available in which parts of the codebase (also touched on in the link above)
- they are _[really slow](https://shopify.engineering/the-case-against-monkey-patching#:~:text=A%20Note%20on%20Refinements)_ (Although there is some data that [disputes](https://www.honeybadger.io/blog/benchmarking-ruby-refinements/) this.)

In the case of the `Hash#assert_valid_keys!`, there's already a similar method being required from `ActiveSupport`, so there's little reason to add a refinement with substantially similar functionality.

### CODE REVIEW

File: Library/Homebrew/extend/hash_validator.rb [hunk 1]
```diff
@@ -1,13 +0,0 @@
-# typed: false
-# frozen_string_literal: true
-
-module HashValidator
-  refine Hash do
-    def assert_valid_keys!(*valid_keys)
```
Comment (dduugg) [inline Library/Homebrew/extend/hash_validator.rb:6]: Note that the `ActiveSupport` implementation [varies](https://github.com/rails/rails/blob/v6.1.7.3/activesupport/lib/active_support/core_ext/hash/keys.rb#L39-L55) a bit, although this shouldn't matter, based on my cursory look at our actual call sites.

**Comment by reitermarkus** [comment]
Based on https://github.com/Homebrew/brew/pull/14928, I thought we wanted to use less ActiveSupport, not more?

**Comment by dduugg** [comment]
> Based on #14928, I thought we wanted to use less ActiveSupport, not more?

Yes, but more specifically, we want to reduce the amount of `ActiveSupport` we _require_, bc it can add a lot of upfront cost, but we currently already have this method available after we load `global.rb`.

If we replace https://github.com/casunlight/rails/blob/master/activesupport/lib/active_support/core_ext/hash/keys.rb we'll likely keep the contents around in `extend` (based on my work with `Array`), so I expect this change to be durable.

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me, thanks again @dduugg

**Pull request merged**